src/cpp/cpp_typecheck_conversions.cpp

```diff
@@ -1880,9 +1874,7 @@ bool cpp_typecheckt::reinterpret_typecast(
 
   if(is_reference(type) && e.get_bool(ID_C_lvalue))
   {
-    exprt tmp(ID_address_of, pointer_typet());
-    tmp.type().subtype()=e.type();
-    tmp.copy_to_operands(e);
+    exprt tmp=address_of_exprt(e, pointer_type(e.type()));
```

Comment:
No, not in this case. Note the following .make_typecast, which destroys the type invariant.